allensdk/brain_observatory/ecephys/stimulus_analysis/drifting_gratings.py

```diff
@@ -0,0 +1,404 @@
+
+        Returns:
+        -------
+        pref_ori - stimulus orientation driving the maximal response
+
+        """
+
+        similar_conditions = [self.stimulus_conditions.index[self.stimulus_conditions[self._col_ori] == ori].tolist() for ori in self.orivals]
+        df = pd.DataFrame(index=self.orivals,
+                         data = {'spike_mean' : 
+                                [self.conditionwise_statistics.loc[unit_id].loc[condition_inds]['spike_mean'].mean() for condition_inds in similar_conditions]
+                             }
+                         ).rename_axis(self._col_ori)
+
+        return df.idxmax().iloc[0]
+
+
+    def _get_pref_tf(self, unit_id):
+
+        """ Calculate the preferred temporal frequency condition for a given unit
+
+        Params:
+        -------
+        unit_id - unique ID for the unit of interest
+
+        Returns:
+        -------
+        pref_tf - stimulus temporal frequency driving the maximal response
+
+        """
+
+        similar_conditions = [self.stimulus_conditions.index[self.stimulus_conditions[self._col_tf] == tf].tolist() for tf in self.tfvals]
+        df = pd.DataFrame(index=self.tfvals,
+                         data = {'spike_mean' : 
+                                [self.conditionwise_statistics.loc[unit_id].loc[condition_inds]['spike_mean'].mean() for condition_inds in similar_conditions]
+                             }
+                         ).rename_axis(self._col_tf)
+
+        return df.idxmax().iloc[0]
+
+
+    def _get_selectivity(self, unit_id, pref_tf, selectivity_type='osi'):
+
+        """ Calculate the orientation or direction selectivity for a given unit
+
+        Params:
+        -------
+        unit_id - unique ID for the unit of interest
+        pref_tf - preferred temporal frequency for this unit
+        selectivity_type - 'osi' or 'dsi'
+
+        Returns:
+        -------
+        selectivity - orientation or direction selectivity value
+
+        """
+        orivals_rad = deg2rad(self.orivals).astype('complex128')
+
+        condition_inds = self.stimulus_conditions[self.stimulus_conditions[self._col_tf] == pref_tf].index.values
+        df = self.conditionwise_statistics.loc[unit_id].loc[condition_inds]
+        df = df.assign(Ori = self.stimulus_conditions.loc[df.index.values][self._col_ori])
+        df = df.sort_values(by=[self._col_ori])
+
+        tuning = np.array(df['spike_mean'].values)
+
+        if selectivity_type == 'osi':
+            return osi(orivals_rad, tuning)
+        elif selectivity_type == 'dsi':
+            return dsi(orivals_rad, tuning)
+
+
+
+    def _get_f1_f0(self, unit_id):
+        """ Calculate F1/F0 for a given unit
+
+        A measure of how tightly locked a unit's firing rate is to the cycles of a drifting grating
+
+        Params:
+        -------
+        unit_id - unique ID for the unit of interest
+
+        Returns:
+        -------
+        f1_f0 - metric
+
+        """
+
+        return np.nan
```

Comment:
could you have these raise an exception? NotImplemented is a bit semantically overloaded, but maybe a subclass or just a sensible message?

Reply:
Should we add a warning instead of raising an exception? If we raise an exception it's going to be tricky to write and test the lism module until these these methods are completed.